### PR TITLE
refactor(pipeline): split scheduler executor

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
split scheduler executor into k8sjob,k8sflink,k8sspark

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=219629&iterationID=1115&pId=290591&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |refactor pipeline scheduler executor               |
| 🇨🇳 中文    |  重构了pipeline的scheduler执行器            |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/pipeline/dbclient/op_pipeline_config.go [hunk 1]
```diff
@@ -39,12 +39,40 @@ var defaultWaitActionExecutor = spec.PipelineConfig{
 	},
 }
 
+var defaultK8SJOBActionExecutor = spec.PipelineConfig{
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline_config.go:42]: K8s
K8sJob

Comment (sfwn) [reply modules/pipeline/dbclient/op_pipeline_config.go:42]: see: https://github.com/kubernetes/kubernetes/blob/master/README.md#kubernetes-k8s

File: modules/pipeline/dbclient/op_pipeline_config.go [hunk 2]
```diff
@@ -39,12 +39,40 @@ var defaultWaitActionExecutor = spec.PipelineConfig{
...
+		Name:    spec.PipelineTaskExecutorNameK8sSparkDefault.String(),
+		Options: nil,
+	},
+}
+
 func (client *Client) ListPipelineConfigsOfActionExecutor() (configs []spec.PipelineConfig, cfgChan chan spec.ActionExecutorConfig, err error) {
 	if err := client.Find(&configs, spec.PipelineConfig{Type: spec.PipelineConfigTypeActionExecutor}); err != nil {
 		return nil, nil, err
 	}
 	// add default api-test action executor
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline_config.go:73]: adjust comment

File: modules/pipeline/pipengine/actionexecutor/logic/errorsx.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/erda-project/erda/modules/pipeline/pkg/task_uuid"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+)
+
+var (
+	errMissingNamespace = errors.New("action missing namespace")
+	errMissingUUID      = errors.New("action missing UUID")
+)
+
+func WrapError(exe string, err *error, op string, action *spec.PipelineTask) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/logic/errorsx.go:31]: Too more args.
Meaningless function name.

File: modules/pipeline/pipengine/actionexecutor/logic/status.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+	case string(apistructs.StatusStoppedOnFailed), string(apistructs.StatusFailed):
+		return apistructs.PipelineStatusFailed
+
+	case string(apistructs.StatusStoppedByKilled):
+		return apistructs.PipelineStatusStopByUser
+
+	case string(apistructs.StatusNotFoundInCluster):
+		// scheduler 返回 job 在 cluster 中不存在 (在 etcd 中存在)，对应为 启动错误
+		// 典型场景：created 成功，env key 为数字，导致 start job 时真正去创建 k8s job 时失败，即启动失败
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/logic/status.go:48]: use English is better.

Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/logic/status.go:48]: and rethink this situation, is right or not.

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 5]
```diff
@@ -37,31 +48,54 @@ func GetManager() *Manager {
...
-		create, ok := m.factory[types.Kind(c.Kind)]
-		if !ok {
-			return errors.Errorf("unregistered action executor kind: %v", c.Kind)
+		create, err := m.GetCreateByKind(types.Kind(c.Kind))
+		if err != nil {
+			return err
+		}
+		kind := types.Kind(c.Kind)
+		if kind.IsK8SKind() {
+			clusters, err := clusterinfo.ListAllClusters()
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/manager.go:65]: Init when truly used, the same with hook on cluster-addition.

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 6]
```diff
@@ -73,11 +107,99 @@ func (m *Manager) Get(name types.Name) (types.ActionExecutor, error) {
...
+	if err != nil {
+		return nil, errors.Errorf("executor [%s] created failed, err: %v", name, err)
+	}
+	m.Lock()
+	m.executors[actionExecutor.Name()] = actionExecutor
+	m.Unlock()
+	return actionExecutor, nil
+}
+
+func (m *Manager) GetCreateByKind(kind types.Kind) (types.CreateFn, error) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/manager.go:138]: CreateFn

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 7]
```diff
@@ -73,11 +107,99 @@ func (m *Manager) Get(name types.Name) (types.ActionExecutor, error) {
...
+	m.Lock()
+	f, ok := m.factory[kind]
+	m.Unlock()
+	if !ok {
+		return nil, errors.Errorf("unregistered action executor kind: [%s]", kind)
+	}
+	return f, nil
+}
+
+func (m *Manager) deleteK8SExecutor(cluster apistructs.ClusterInfo) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/manager.go:151]: split to a new file for k8s related functions.

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go [hunk 8]
```diff
@@ -66,20 +75,90 @@ func (k *K8sFlink) Status(ctx context.Context, task *spec.PipelineTask) (apistru
...
-	statusDesc = composeStatusDesc(flinkCluster.Status)
+	status := composeStatusDesc(flinkCluster.Status)
+	statusDesc.Status = logic.TransferStatus(string(status.Status))
+	statusDesc.Desc = status.Reason
 
 	return statusDesc, nil
 }
 
-func (k *K8sFlink) Create(ctx context.Context, task *spec.PipelineTask) (interface{}, error) {
+func (k *K8sFlink) Exist(ctx context.Context, task *spec.PipelineTask) (created bool, started bool, err error) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go:93]: You copied these related code three times. It's extremely wrong.
You should add these logics to a common pkg.

File: modules/pipeline/pipengine/actionexecutor/types/define.go [hunk 9]
```diff
@@ -72,6 +73,52 @@ func (s Kind) Validate() bool {
 	return formatter.MatchString(string(s))
 }
 
+func (s Kind) IsK8SKind() bool {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/types/define.go:76]: add unit-test.

File: modules/pipeline/services/pipelinesvc/create_task.go [hunk 10]
```diff
@@ -183,14 +182,30 @@ func (s *PipelineSvc) calculateTaskRunAfter(action *pipelineyml.Action) []string
 }
 
 // judgeTaskExecutor judge task executor by action info
-func (s *PipelineSvc) judgeTaskExecutor(action *pipelineyml.Action, actionSpec *apistructs.ActionSpec) (spec.PipelineTaskExecutorKind, spec.PipelineTaskExecutorName) {
+func (s *PipelineSvc) judgeTaskExecutor(task *spec.PipelineTask, actionSpec *apistructs.ActionSpec) (spec.PipelineTaskExecutorKind, spec.PipelineTaskExecutorName) {
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/create_task.go:185]: add ut

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/dbclient/op_pipeline_config.go (hunk 1 above)
Comment (chengjoey) [reply modules/pipeline/dbclient/op_pipeline_config.go:42]: adjusted

File: modules/pipeline/pipengine/actionexecutor/logic/errorsx.go (hunk 3 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/errorsx.go:31]: adjusted

File: modules/pipeline/pipengine/actionexecutor/logic/status.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/status.go:48]: used english, maybe notfound in cluster status shouldn't be start error

File: modules/pipeline/pipengine/actionexecutor/manager.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/manager.go:138]: adjusted

File: modules/pipeline/pipengine/actionexecutor/manager.go (hunk 7 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/manager.go:151]: adjusted

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go (hunk 8 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go:93]: adjusted, add these logics to logic.status file

File: modules/pipeline/pipengine/actionexecutor/types/define.go (hunk 9 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/types/define.go:76]: added

File: modules/pipeline/services/pipelinesvc/create_task.go (hunk 10 above)
Comment (chengjoey) [reply modules/pipeline/services/pipelinesvc/create_task.go:185]: already existed in pipelinesvc.create_task_test.go file

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 11]
```diff
@@ -73,11 +95,54 @@ func (m *Manager) Get(name types.Name) (types.ActionExecutor, error) {
...
+		return e, nil
+	}
+	m.RLock()
+	kind, ok := m.kinds[name]
+	m.RUnlock()
+	if !ok {
+		// if the executor is not found, it may be a k8s type executor
+		// so try to get kind from name and store in the kinds cache
+		var found bool
+		kind, found = m.GetK8sKindByExecutorName(name)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/manager.go:111]: check k8s kind here is also bad.

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go [hunk 12]
```diff
@@ -66,20 +75,72 @@ func (k *K8sFlink) Status(ctx context.Context, task *spec.PipelineTask) (apistru
...
+	defer logic.WrapTaskErrorForExecutor(k.Kind().String(), &err, "create job", task)
+	if err := logic.ValidateAction(task); err != nil {
+		return nil, err
+	}
+	created, _, err := k.Exist(ctx, task)
+	if err != nil {
+		return nil, err
+	}
+	if created {
+		logrus.Warnf("%s: task already created, actionInfo: %s", k.Kind(), logic.PrintActionInfo(task))
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go:117]: task info

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sjob/k8sjob.go [hunk 13]
```diff
@@ -128,26 +137,79 @@ func (k *K8sJob) Status(ctx context.Context, action *spec.PipelineTask) (desc ap
...
+			err = nil
+			return
+		}
+		return
+	}
+	return logic.JudgeExistedByStatus(statusDesc)
+}
+
+func (k *K8sJob) Create(ctx context.Context, task *spec.PipelineTask) (data interface{}, err error) {
+	defer logic.WrapTaskErrorForExecutor(k.Kind().String(), &err, "create job", task)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/k8sjob/k8sjob.go:177]: you should have a default implement for k8s kind.
and overwrite necessary methods.

File: modules/pipeline/pipengine/actionexecutor/types/define.go (hunk 9 above)
Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/types/define.go:76]: K8s

File: modules/pipeline/spec/pipeline_configs.go [hunk 14]
```diff
@@ -30,6 +30,12 @@ type ActionExecutorConfig struct {
 	Options map[string]string `json:"options,omitempty"`
 }
 
+func (a *ActionExecutorConfig) IsK8SKind() bool {
```
Comment (sfwn) [inline modules/pipeline/spec/pipeline_configs.go:33]: so many IsK8sKind methods.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/actionexecutor/logic/status.go [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+		return apistructs.PipelineStatusSuccess
+
+	case string(apistructs.StatusStoppedOnFailed), string(apistructs.StatusFailed):
+		return apistructs.PipelineStatusFailed
+
+	case string(apistructs.StatusStoppedByKilled):
+		return apistructs.PipelineStatusStopByUser
+
+	case string(apistructs.StatusNotFoundInCluster):
+		// k8sjob, k8sspark, k8sflink return start error status if not found in cluster
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/status.go:51]: this  status is for compatibility with scheduler executor
func-exist will transfer to started false

File: modules/pipeline/dbclient/op_pipeline_config.go [hunk 16]
```diff
@@ -39,12 +39,40 @@ var defaultWaitActionExecutor = spec.PipelineConfig{
 	},
 }
 
+var defaultK8sJOBActionExecutor = spec.PipelineConfig{
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/dbclient/op_pipeline_config.go:42]: adjusted

File: modules/pipeline/pipengine/actionexecutor/k8s_helper.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+	defer m.RUnlock()
+	for k := range m.factory {
+		if k.IsK8SKind() && strings.HasPrefix(executorName.String(), k.MakeK8SKindExecutorName("").String()) {
+			return k, true
+		}
+	}
+	return "", false
+}
+
+func (m *Manager) deleteK8SExecutor(cluster apistructs.ClusterInfo) {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/k8s_helper.go:42]: adjusted

File: modules/pipeline/pipengine/actionexecutor/k8s_helper.go [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/conf"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/actionexecutor/types"
+	"github.com/erda-project/erda/modules/pipeline/pkg/clusterinfo"
+)
+
+func (m *Manager) GetK8sKindByExecutorName(executorName types.Name) (types.Kind, bool) {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/k8s_helper.go:31]: adjusted, impl the GetKindByExecutorName method, this method will return all kind

File: modules/pipeline/pipengine/actionexecutor/k8s_helper_test.go [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+
+func TestGetK8sKindByExecutorName(t *testing.T) {
+	testCases := []struct {
+		name          types.Name
+		expectedKind  types.Kind
+		expectedFound bool
+	}{
+		{
+			name:          types.Name("apitest"),
+			expectedKind:  "",
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/k8s_helper_test.go:34]: adjusted

File: modules/pipeline/pipengine/actionexecutor/logic/errorsx.go [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/erda-project/erda/modules/pipeline/pkg/task_uuid"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+)
+
+var (
+	errMissingNamespace = errors.New("action missing namespace")
+	errMissingUUID      = errors.New("action missing UUID")
+)
+
+func WrapTaskErrorForExecutor(exe string, err *error, op string, action *spec.PipelineTask) {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/errorsx.go:31]: adjusted, k8s type executor will init the err wrapper
image

File: modules/pipeline/pipengine/actionexecutor/logic/status.go [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+		return apistructs.PipelineStatusStartError
+	}
+
+	return apistructs.PipelineStatusUnknown
+}
+
+func JudgeExistedByStatus(statusDesc apistructs.PipelineStatusDesc) (created bool, started bool, err error) {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/status.go:58]: added

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 22]
```diff
@@ -15,17 +15,26 @@
...
 	"github.com/erda-project/erda/modules/pipeline/spec"
+	"github.com/erda-project/erda/pkg/goroutinepool"
 )
 
 // Manager is an executor manager, it holds the all executor instances.
 type Manager struct {
+	sync.RWMutex
 	factory   map[types.Kind]types.CreateFn
 	executors map[types.Name]types.ActionExecutor
+	kinds     map[types.Name]types.Kind
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/manager.go:36]: adjusted

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 23]
```diff
@@ -37,31 +46,44 @@ func GetManager() *Manager {
...
+		m.Lock()
+		m.kinds[name] = kind
+		m.executors[actionExecutor.Name()] = actionExecutor
+		m.Unlock()
 
-		m.executors[name] = actionExecutor
-		logrus.Infof("=> kind [%s], name [%s], created", c.Kind, c.Name)
+		logrus.Infof("=> kind [%s] created", c.Kind)
 	}
+	triggerChan := clusterinfo.RegisterRefreshChan()
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/manager.go:84]: move to ListenAndPatchK8sExecutor func
image

File: modules/pipeline/pipengine/actionexecutor/manager.go (hunk 11 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/manager.go:111]: adjusted, use `GetKindByExecutorName` get all kind
image

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go (hunk 12 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go:117]: adjusted

File: modules/pipeline/spec/pipeline_configs.go (hunk 14 above)
Comment (chengjoey) [reply modules/pipeline/spec/pipeline_configs.go:33]: two place

File: modules/pipeline/pipengine/actionexecutor/manager.go [hunk 24]
```diff
@@ -36,32 +45,43 @@ func GetManager() *Manager {
...
+		m.Lock()
+		m.kindsByName[name] = kind
+		m.executorsByName[actionExecutor.Name()] = actionExecutor
+		m.Unlock()
 
-		m.executors[name] = actionExecutor
-		logrus.Infof("=> kind [%s], name [%s], created", c.Kind, c.Name)
+		logrus.Infof("=> kind [%s] created", c.Kind)
 	}
+	go m.ListenAndPatchK8sExecutor(context.Background())
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/manager.go:84]: use the context passed in by Initialize.

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go [hunk 25]
```diff
@@ -66,20 +75,72 @@ func (k *K8sFlink) Status(ctx context.Context, task *spec.PipelineTask) (apistru
...
+	if err != nil {
+		return nil, err
+	}
+	if !created {
+		logrus.Warnf("%s: task not created, try to create actionInfo: %s", k.Kind().String(), logic.PrintTaskInfo(task))
+		_, err = k.Create(ctx, task)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Warnf("k8sjob: action created, continue to start, actionInfo: %s", logic.PrintTaskInfo(task))
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go:137]: bad k8sjob in k8sflink.

Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go:137]: check all logs pls

File: modules/pipeline/pipengine/actionexecutor/types/define.go [hunk 26]
```diff
@@ -72,6 +73,52 @@ func (s Kind) Validate() bool {
...
 
+func (s Kind) IsK8sKind() bool {
+	switch s {
+	case Kind(spec.PipelineTaskExecutorKindK8sJob), Kind(spec.PipelineTaskExecutorKindK8sFlink), Kind(spec.PipelineTaskExecutorKindK8sSpark):
+		return true
+	}
+	return false
+}
+
+func (s Kind) MakeK8SKindExecutorName(clusterName string) Name {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/types/define.go:84]: K8s

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/actionexecutor/types/define.go (hunk 26 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/types/define.go]: ad

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go (hunk 25 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sflink/flink.go]: ad

File: modules/pipeline/pipengine/actionexecutor/manager.go (hunk 24 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/manager.go]: ad

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sjob/k8sjob.go (hunk 13 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sjob/k8sjob.go]: define the k8s base executor interface
impl the `exist,create,remove,batchdelete` for k8s kind
image

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go [hunk 27]
```diff
@@ -45,16 +47,40 @@ const (
...
+	defer k.errWrapper.WrapTaskError(&err, "start job", task)
+	if err := logic.ValidateAction(task); err != nil {
+		return nil, err
+	}
+	created, started, err := k.Exist(ctx, task)
+	if err != nil {
+		return nil, err
+	}
+	if !created {
+		logrus.Warnf("%s: task not created, try to create actionInfo: %s", k.Kind().String(), logic.PrintTaskInfo(task))
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go:73]: what is `try to create actionInfo`?

File: modules/pipeline/pkg/containers/task_container.go [hunk 28]
```diff
@@ -15,25 +15,22 @@
...
 	"github.com/erda-project/erda/modules/pipeline/spec"
 )
 
 func GenContainers(task *spec.PipelineTask) ([]apistructs.TaskContainer, error) {
-	if value, ok := task.Extra.Action.Params["bigDataConf"]; ok {
-		spec := apistructs.BigdataSpec{}
-		if err := json.Unmarshal([]byte(value.(string)), &spec); err != nil {
-			return nil, fmt.Errorf("failed to unmarshal task bigDataConf")
-		}
+	if spec, err := task.GetBigDataConf(); err == nil {
```
Comment (sfwn) [inline modules/pipeline/pkg/containers/task_container.go:25]: bad pattern. you split the two normal situations (flink+spark and task).
return as soon as possible if err occurred.
then check spec fields.

File: modules/pipeline/providers/reconciler/legacy/reconciler/taskrun/taskop/prepare.go [hunk 29]
```diff
@@ -357,7 +356,9 @@ func (pre *prepare) makeTaskRun() (needRetry bool, err error) {
 
 	if p.Extra.StorageConfig.EnableNFSVolume() &&
 		!p.Extra.StorageConfig.EnableShareVolume() &&
-		task.ExecutorKind == spec.PipelineTaskExecutorKindScheduler {
+		task.ExecutorKind == spec.PipelineTaskExecutorKindK8sJob ||
```
Comment (sfwn) [inline modules/pipeline/providers/reconciler/legacy/reconciler/taskrun/taskop/prepare.go:359]: can you use `IsK8sKind` here?

File: modules/pipeline/services/pipelinesvc/create_task.go [hunk 30]
```diff
@@ -53,9 +53,10 @@ func (s *PipelineSvc) makeNormalPipelineTask(p *spec.Pipeline, ps *spec.Pipeline
 	// task.Extra.Envs
 	// task.Extra.Labels
 	// task.Extra.Image
+	task.Extra.Action = *action
 
 	// set executor
-	executorKind, executorName := s.judgeTaskExecutor(action, passedDataWhenCreate.GetActionJobSpecs(extmarketsvc.MakeActionTypeVersion(action)))
+	executorKind, executorName := s.judgeTaskExecutor(task, passedDataWhenCreate.GetActionJobSpecs(extmarketsvc.MakeActionTypeVersion(action)))
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/create_task.go:59]: you have to test that running-tasks with old `SCHEDULER` executor.
does these tasks can be continuous reconcile after this pr merged?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/services/pipelinesvc/create_task.go (hunk 30 above)
Comment (chengjoey) [reply modules/pipeline/services/pipelinesvc/create_task.go:59]: impl `GetExecutorName` func, this func is compatible with old scheduler executor
image
image

File: modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go (hunk 27 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/k8sspark/k8sspark.go:73]: adjusted, k8sjob, k8sflink k8sspark had change to `%s: task not created(auto try to create), taskInfo: %s`

File: modules/pipeline/pkg/containers/task_container.go (hunk 28 above)
Comment (chengjoey) [reply modules/pipeline/pkg/containers/task_container.go]: adjusted

File: modules/pipeline/providers/reconciler/legacy/reconciler/taskrun/taskop/prepare.go (hunk 29 above)
Comment (chengjoey) [reply modules/pipeline/providers/reconciler/legacy/reconciler/taskrun/taskop/prepare.go]: adjusted

File: modules/pipeline/services/pipelinesvc/create_task.go (hunk 30 above)
Comment (chengjoey) [reply modules/pipeline/services/pipelinesvc/create_task.go:59]: added ut in pipeline_task.go

Comment (chengjoey) [reply modules/pipeline/services/pipelinesvc/create_task.go:59]: tested, both old scheduler and new k8sjob is success
image

**Review by sfwn — APPROVED** [review]

**Pull request merged**